src/charm.py

```diff
@@ -940,569 +472,497 @@ def _handle_status_update_config(self, event: HookEvent) -> None:
             return
 
         self._update_kratos_external_idp_configurations()
-        self._cleanup_peer_data()
-        self.cert_transfer.push_ca_certs()
-        self._update_config()
-        # We need to push the layer because this may run before _on_pebble_ready
-        self._container.add_layer(WORKLOAD_CONTAINER_NAME, self._pebble_layer, combine=True)
+
+        self._workload_service.push_ca_certs(
+            TLSCertificates.load(self.certificate_transfer_requirer).ca_bundle
+        )
+
+        if template := self.charm_config["recovery_email_template"]:
+            self._container.push(EMAIL_TEMPLATE_FILE_PATH, template, make_dirs=True)
+
+        identity_schema_cm = IdentitySchemaConfigMap(
+            self._k8s_client, self.model.name, self.app.name
+        )
+        identity_schema = IdentitySchema(
+            [
+                CharmConfigIdentitySchemaProvider(self.charm_config),
+                ConfigMapIdentitySchemaProvider(identity_schema_cm),
+                DefaultIdentitySchemaProvider(),
+            ],
+        )
+
+        config_file = ConfigFile.from_sources(
+            self.charm_config,
+            ClaimMapper(),
+            LoginUIEndpointData.load(self.login_ui_requirer),
+            PublicIngressData.load(self.public_ingress),
+            RegistrationWebhookData.load(self.registration_webhook_requirer),
+            ExternalIdpIntegratorData.load(self.external_idp_requirer),
+            identity_schema,
+            OIDCProviderConfigMap(self._k8s_client, self.model.name, self.app.name),
+        )
+
         try:
-            self._restart_service(restart=self.config_changed)
-        except ChangeError as err:
-            logger.error(str(err))
+            self._pebble_service.plan(self._pebble_layer, config_file)
+        except PebbleServiceError as e:
+            logger.error(f"Failed to start the service, please check the container logs: {e}")
```

Comment:
nit: we use both f-strings and printf style for logs, shall we stick to one approach?